agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1504 @@
+                    k8s::try_delete_instance_arc(
+                        kube_interface.clone(),
+                        instance.clone(),
+                        self.config.metadata.namespace.as_ref().unwrap().clone(),
+                    )
+                    .await
+                    .unwrap();
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub mod start_discovery {
+    use super::super::registration::{
+        DiscoveryHandlerConnectivityStatus, DiscoveryHandlerDetails, DiscoveryHandlerEndpoint,
+    };
+    // Use this `mockall` macro to automate importing a mock type in test mode, or a real type otherwise.
+    #[double]
+    pub use super::DiscoveryOperator;
+    use super::{try_receive, StreamType};
+    use akri_shared::k8s;
+    use mockall_double::double;
+    use std::{sync::Arc, time::Duration};
+    use tokio::sync::broadcast;
+
+    /// This is spawned as a task for each Configuration and continues to run
+    /// until the Configuration is deleted, at which point, this function is signaled to stop.
+    /// It consists of three subtasks:
+    /// 1) Initiates discovery on all already registered discovery handlers in the RegisteredDiscoveryHandlerMap
+    /// with the same protocol name as the Configuration (Configuration.protocol.name).
+    /// 2) Listens for new discover handlers to come online for this Configuration and initiates discovery.
+    /// 3) Checks whether Offline Instances have exceeded their grace period, in which case it
+    /// deletes the Instance.
+    pub async fn start_discovery(
+        discovery_operator: DiscoveryOperator,
+        new_discovery_handler_sender: broadcast::Sender<String>,
+        stop_all_discovery_sender: broadcast::Sender<()>,
+        finished_all_discovery_sender: &mut broadcast::Sender<()>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let config = discovery_operator.get_config();
+        info!(
+            "start_discovery - entered for protocol {}",
+            config.spec.protocol.name
+        );
+        let config_name = config.metadata.name.clone();
+        let mut tasks = Vec::new();
+        let discovery_operator = Arc::new(discovery_operator);
+
+        // Call discover on already registered Discovery Handlers for this Configuration's protocol
+        let task1_discovery_operator = discovery_operator.clone();
+        tasks.push(tokio::spawn(async move {
+            do_discover(
+                task1_discovery_operator,
+                Arc::new(Box::new(k8s::create_kube_interface())),
+            )
+            .await
+            .unwrap();
+        }));
+
+        // Listen for new discovery handlers to call discover on
+        let mut stop_all_discovery_receiver = stop_all_discovery_sender.subscribe();
+        let mut new_discovery_handler_receiver = new_discovery_handler_sender.subscribe();
+        let task2_discovery_operator = discovery_operator.clone();
```

Comment:
maybe more descriptive name: `new_dh_discovery_operator`

Reply:
fixed in 88b4c709